### ISSUE TITLE
Inconsistency when using mathematical operations on copies of tensors

### ISSUE THREAD

**Issue by larceci89** [google/TensorNetwork#490]
In the setting I specify below, I found inconsistent results when using 
1 ---  node.tensor = node.tensor / num
2 ---  node.tensor /= num
which I suppose should give the same result, for some number num.

In the code below, I define a list of two nodes and give it in input to two almost identical subroutines, whose only difference is on how they perform the division by 1000 on one tensor. 
I expect the node given in input to remain untouched by the routine. Indeed, I copy the input nodes list and perform computations just on that copy. Routine 1 fulfills this expectation, while routine 2 changes my input nodes, although I acted on copies of them.

Let me paste here the example code I was referring to:
#------------------------------------------------------------------------------
```python
import numpy as np
import tensornetwork as tn

def copy_nodeslist(nlst): 
	ndict, _ = tn.copy( nlst )
	newlist = [ ndict[tens] for tens in nlst ]
	return newlist

def routine1( nodes_inpt ):
	newnodes = copy_nodeslist( nodes_inpt )
	newnodes[0].tensor = newnodes[0].tensor / 1000
	return newnodes

def routine2( nodes_inpt ):
	newnodes = copy_nodeslist( nodes_inpt )
	newnodes[0].tensor /= 1000
	return newnodes

nodes = [ tn.Node( np.random.randn(2,2), name='T1' ), tn.Node( np.random.randn(2,2), name='T2' ) ]
print('Tensor before entering subroutines:\n{}'.format( nodes[0].tensor ))

newnode = routine1( nodes )
print('Tensor after routine 1:\n{}'.format( nodes[0].tensor ))

newnode = routine2( nodes )
print('Tensor after routine 2:\n{}'.format( nodes[0].tensor ))
```
#------------------------------------------------------------------------------

**Comment by Thenerdstation** [comment]
I believe the issue is around our `set_tensor` operation. Let me take a look.

**Comment by Thenerdstation** [comment]
Aw, yes I see the issue now. This is an issue with numpy arrays be mutable.

So what is happening when you do `tn.copy(nodes)` is that all of the structure of the nodes is replicated and the _references_ to the original arrays is used for the `node.tensor`. There we a few reasons for this.

1. You need the reference to the original node for gradient calculations. It's pretty common to copy-then-contract a network for some machine learning applications.
2. for JAX (which is what we use on our team), the tensors are totally immutable by design, so replicating a reference allows us to save memory.

However, I can see why this is super confusing when using numpy. We would be able to fix this by adding a new `replicate_tensor` method to the `BaseBackend` class and have it return a "copy" however that is efficiently defined for your backend.

**Comment by Thenerdstation** [comment]
And as you can see, this issue doesn't happen if you use JAX instead.

```python
import jax.numpy as np
import numpy as onp
import tensornetwork as tn
tn.set_default_backend("jax")

def copy_nodeslist(nlst): 
	ndict, _ = tn.copy( nlst )
	newlist = [ ndict[tens] for tens in nlst ]
	return newlist

def routine1( nodes_inpt ):
  newnodes = copy_nodeslist( nodes_inpt )
  val = newnodes[0].tensor / 1000
  newnodes[0].tensor = val
  return newnodes

def routine2( nodes_inpt ):
	newnodes = copy_nodeslist( nodes_inpt )
	newnodes[0].tensor /= 1000
	return newnodes

nodes = [ tn.Node( onp.random.randn(2,2), name='T1' ), tn.Node( onp.random.randn(2,2), name='T2' ) ]
print('Tensor before entering subroutines:\n{}'.format( nodes[0].tensor ))

newnode = routine1( nodes )
print('Tensor after routine 1:\n{}'.format( nodes[0].tensor ))

newnode = routine2( nodes )
print('Tensor after routine 2:\n{}'.format( nodes[0].tensor ))
```

Outputs:
```
Tensor before entering subroutines:
[[ 0.38292316  0.12522404]
 [-1.0760983  -0.20792884]]
Tensor after routine 1:
[[ 0.38292316  0.12522404]
 [-1.0760983  -0.20792884]]
Tensor after routine 2:
[[ 0.38292316  0.12522404]
 [-1.0760983  -0.20792884]]
```

**Comment by larceci89** [comment]
Ah I see, it was a subtle issue indeed! Thank you for the reply! Then I will switch to the jax backend to reduce the probability of finding other similar issues.
